### PR TITLE
Trans-Q Phase II

### PR DESCRIPTION
**Pull request by thomas-niehaus**
Further work to bring RangeSep routines back into main TD-DFTB routines.

- Removal of calls to indxoo and indxvv where possible 
- Removal of direct calls to transQ
- Stratmann diagonalizer as alternative to Arpack for standard TD-DFTB. Works for S/T, spin pol, temperature, windowing. Not yet for onsite corrections and not yet OpenMP

Stratmann is significantly faster for small nExc (factor 2-10), but Arpack is faster for nExc around 100. Stratmann has accuracy issues for osc strength (deviation 0.01 %) if initial SubSpace is too small.

TODO in Phase III
- Implementation of rangesep in actionAplusB
- Addition of rangesep to forces
- Removal of old rangesep routines
- Full removal of indxov and direct calls to transQ
- Removal of updwn 
- Derived type for linresp parameters
- Removal of need for ARPACK in make

### CODE REVIEW

File: test/prog/dftb+/timedep/C4H6-S1-Force_uncache-Stratmann/XREST.DAT [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Mulliken analysis of excited state S 1
```
Comment (bhourahine) [inline test/prog/dftb+/timedep/C4H6-S1-Force_uncache-Stratmann/XREST.DAT:1]: Any reason to check this file in, as its not used by the regression testing? (Likewise for other .DAT files)

Comment (thomas-niehaus) [reply test/prog/dftb+/timedep/C4H6-S1-Force_uncache-Stratmann/XREST.DAT:1]: No, happened by mistake. Will remove it.

**Comment by aradi** [comment]
@thomas-niehaus Do you have any other branches derived from this one? I'd like to try to undo the last two commits here (merge with main and removing redundant files) and doing only latter. We would have to squash the commits in order to really remove the redundant files from the history. But that only works properly, if no merges from other branches (e.g. main) are present.

**Comment by thomas-niehaus** [comment]
No, I have no derived branches. Please go ahead!

**Comment by aradi** [comment]
OK, I've removed the master-merge from this brach, so that we can easily squash commits later. You would have to update your local copy as well:
```
git checkout master
git branch -D transq-PhaseII
git fetch origin transq-PhaseII
git checkout transq-PhaseII
```
(provided you call your GitHub repository `origin`)

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 2]
```diff
@@ -535,66 +606,285 @@ subroutine omegatvec(spin, vin, vout, wij, sym, win, nmatup, iAtomStart, stimc,
...
+          & grndEigVecs, onsMEs, orb, vTmp, vout_ons)
       vout(:) = vout + vout_ons
     end if
 
-    vout(:) = vout(:) + ( wij**2 ) * vin
+    vout(:) = vout(:) + wij(:) * vTmp(:)
+
+    if(.not. tAplusB) then
+       vout(:) = vout(:) * sqrt(wij(:))
+    endif
```
Comment (aradi) [inline prog/dftb+/lib_timedep/linrespcommon.F90:654]: omit `(:)` on RHS

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 3]
```diff
@@ -535,66 +606,285 @@ subroutine omegatvec(spin, vin, vout, wij, sym, win, nmatup, iAtomStart, stimc,
...
+
+  end subroutine actionAminusB
+
+  !> Generates initial matrices M+ and M- for the RPA algorithm by Stratmann
+  !> (JCP 109 8218 (1998).
+  !> M+/- = (A+/-B)_ias,jbt (spin polarized) (A+/-B)^{S/T}_ia,jb (closed shell) 
+  !> Here ias,jbt <= initDim
+  !> Also computed is v+/- = (A+/-B)_ias,jbt with ias <= nMat, jbt <= initDim
+  !> Note: Routine not set up to handle onsite corrections.
+  !> Note: Not yet OpenMP parallelized
```
Comment (aradi) [inline prog/dftb+/lib_timedep/linrespcommon.F90:705]: As it calls subroutines with OMP parallelisation inside (e.g. `hemv`), probably this routine should not be OpenMP parallelized in its current form anyway...

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 4]
```diff
@@ -415,45 +420,101 @@ subroutine wtdn(wij, occNr, win, nmatup, nmat, getia, wn_ij)
 
   end subroutine wtdn
 
+  !> Square root of differences in occupations between filled and empty states
+  !> We need occupations per spin channel ([0:1]) also for closed shell systems
+  subroutine wtdn_tmp(occNr, win, nmatup, nmat, getia, tSpin, n_ij)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:425]: This routine will be removed anyway in the next PR.

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 5]
```diff
@@ -535,66 +606,285 @@ subroutine omegatvec(spin, vin, vout, wij, sym, win, nmatup, iAtomStart, stimc,
...
 
-  end subroutine omegatvec
+    integer :: nMat
 
+    nMat = size(vin) 
+    vOut(:) = 0.0_dp
+
+    ! orb. energy difference diagonal contribution
+    vOut(:) = vOut(:) + wIA(:) * vIn(:) 
+
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:696]: vOut = wIA *  wIn will work at this stage, but not if RS is added.

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 6]
```diff
@@ -535,66 +606,285 @@ subroutine omegatvec(spin, vin, vout, wij, sym, win, nmatup, iAtomStart, stimc,
...
+      do jj = ii, initDim
+        mP(ii,jj) = vP(ii,jj)
+        mP(jj,ii) = mP(ii,jj)
+        mM(ii,jj) = vM(ii,jj)
+        mM(jj,ii) = mM(ii,jj)
+      end do
+    end do
+
+    deallocate(oTmp)
+    deallocate(qTr)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:877]: ok!

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 7]
```diff
@@ -1233,4 +1523,168 @@ subroutine writeExcMulliken(sym, nstat, dq, dqex, coord0, fdMulliken)
 
   end subroutine writeExcMulliken
 
+  !> Set value of sizeIn to new, three times as large, value
+  pure subroutine TN_incSizeInt(sizeIn)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:1527]: Good point.

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 8]
```diff
@@ -1233,4 +1523,168 @@ subroutine writeExcMulliken(sym, nstat, dq, dqex, coord0, fdMulliken)
...
+
+    integer, intent(inout) :: sizeIn
+    real(dp), allocatable, intent(inout) :: vec(:,:)
+
+    integer :: dim1
+    real(dp), allocatable :: temp(:,:)
+    dim1 = size(vec, dim=1)
+    allocate(temp(dim1, 3 * sizeIn))
+    temp(:,:) = 0.0_dp
+    ! Check: would be nice if temp could become memory for vec, see if possible in fortran
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:1546]: ok

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 9]
```diff
@@ -1233,4 +1523,168 @@ subroutine writeExcMulliken(sym, nstat, dq, dqex, coord0, fdMulliken)
...
+
+  !>increase size of a sizeInxsizeIn array
+  pure subroutine TN_incSizeMatBoth(sizeIn, vec)
+
+    integer, intent(inout) :: sizeIn
+    real(dp), allocatable, intent(inout) :: vec(:,:)
+
+    real(dp), allocatable :: temp(:,:)
+
+    allocate(temp(3 * sizeIn, 2 * sizeIn))
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:1595]: Good question, Julian inserted them.

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 10]
```diff
@@ -1233,4 +1523,168 @@ subroutine writeExcMulliken(sym, nstat, dq, dqex, coord0, fdMulliken)
...
+    temp(:,:) = 0.0_dp
+    ! Check: would be nice if temo could become memory for vec, see if possible in fortran
+    temp(1:sizeIn, 1:sizeIn) = vec
+    call move_alloc(temp, vec)
+
+  end subroutine TN_incSizeMatBoth
+
+  !> Same routine exists in rs_linresp and will be removed 
+  !> Calculate square root and inverse of sqrt of a real, symmetric positive definite matrix
+  subroutine calcMatrixSqrt_TN(matIn, spaceDim, memDim, workArray, workDim, matOut, matInvOut)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:1605]: See above, will be removed in next PR.

File: prog/dftb+/lib_dftbplus/initprogram.F90 [hunk 11]
```diff
@@ -5068,6 +5077,16 @@ subroutine ensureLinRespConditions(tSccCalc, t3rd, tRealHS, tPeriodic, tCasidaFo
       end if
     end if
 
+    if (input%ctrl%lrespini%tUseArpack) then
+      if (input%ctrl%lrespini%subSpaceFactorStratmann /= 20) then
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_dftbplus/initprogram.F90:5081]: Yes, I agree. But how to solve this in a more elegant way? I could define a logical isSubSpaceStratmannSet in the parser, but that is overkill.

File: prog/dftb+/lib_dftbplus/parser.F90 [hunk 12]
```diff
@@ -4654,6 +4654,8 @@ subroutine readExcited(node, geo, ctrl)
       end if
       call getChildValue(child, "WriteStatusArnoldi", ctrl%lrespini%tArnoldi, default=.false.)
       call getChildValue(child, "TestArnoldi", ctrl%lrespini%tDiagnoseArnoldi, default=.false.)
+      call getChildValue(child, "UseArpack", ctrl%lrespini%tUseArpack, default=.true.)
+      call getChildValue(child, "SubSpaceStratmann", ctrl%lrespini%subSpaceFactorStratmann, 20)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_dftbplus/parser.F90:4658]: Yes, sounds good, Let me see if I can implement that.

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 13]
```diff
@@ -415,45 +420,101 @@ subroutine wtdn(wij, occNr, win, nmatup, nmat, getia, wn_ij)
...
+    logical, intent(in) :: tSpin
+
+    !> resulting scaled matrix
+    real(dp), intent(out) :: n_ij(:)
+
+    integer :: ia, ii, jj, ss
+    logical :: updwn
+    real(dp) :: docc_ij
+
+    !$OMP PARALLEL DO DEFAULT(SHARED) PRIVATE(ia, ii, jj, updwn, docc_ij) SCHEDULE(RUNTIME)
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:452]: On a different note: Did you see any effect of OpenMP for the example jobs I sent you?

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 14]
```diff
@@ -498,33 +563,39 @@ subroutine omegatvec(spin, vin, vout, wij, sym, win, nmatup, iAtomStart, stimc,
...
-    call wtdn(wij, occNr, win, nmatup, nmat, getia, wnij)
-    wnij = sqrt(wnij) ! always used as root(wnij) after this
+    ! Compute sqrt(n_as-n_is)
+    call wtdn_tmp(occNr, win, nmatup, nmat, getia, spin, sqrOccIA)
+ 
+    if(tAplusB) then
+       vTmp(:) = vin(:) 
+    else
+       vTmp(:) = vin(:) * sqrt(wij(:))
+    endif
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:574]: ok

File: prog/dftb+/lib_timedep/linrespcommon.F90 [hunk 15]
```diff
@@ -474,22 +535,26 @@ subroutine omegatvec(spin, vin, vout, wij, sym, win, nmatup, iAtomStart, stimc,
...
     !> machinery for transition charges between single particle levels
     type(TTransCharges), intent(in) :: transChrg
 
-    integer :: nmat, natom, ia
+    integer :: nmat, natom, ia, ss
 
+    ! This OMP directive seems strange, no ii,jj used. 
   #:if WITH_OMP
     integer :: ii, jj
   #:endif
```
[the remark this answers is not in the document]
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:549]: ok

File: prog/dftb+/lib_timedep/linrespcommon.F90 (hunk 3 above)
Comment (thomas-niehaus) [reply prog/dftb+/lib_timedep/linrespcommon.F90:705]: ok

**Review by thomas-niehaus — COMMENTED** [review]
!! TestArnoldi and WriteStatusArnoldi still appears in dftb_pin.hsd
 Not sure how to apply removeChild correctly

**Comment by aradi** [comment]
The problem is the wrong handling of the `Diagonalizer` keyword (which leads to all the test failures in Debug mode). It is easy to fix, the only questions is, how we should handle the default? Should it be ARPACK, if the code was compiled with ARPACK. Should it be Stratman? (To make sure, it also works if code was built without ARPACK)? Or should we force the user to an explicit specification of the diagonalizer in each case? (Maybe not very user friendly). I am leaning towards Stratman as default at the moment, unless there are any objections from your side.

**Comment by thomas-niehaus** [comment]
Pro Stratmann:
- Much faster for low number of excited states
- Needs no external libs
- Only way to do TD-LC-DFTB
- Better chance to render full MPI 
Pro Arpack
- Well tested over time (little experience with Stratmann, making Stratmann default is somewhat risky)
- Better perf for large number of states
- All modes (including onsite) work
- Better runtime info

I have no clear winner here.
- Backward compatibility (small diff in osc strengths for Stratmann if Subspace too small)

**Comment by aradi** [comment]
I've fixed the parser. However, it can not merged yet, as the Stratman uses uninitialized elements. 

@thomas-niehaus Could you have a look at that? If you compile with the NAG compiler (use `-DCMAKE_BUILD_TYPE=Debug -DWITH_OMP=0`) on dpdevel and execute the timedep/C4H6-Singlet-Stratmann test, you get `NaN` for the oscillator strength array in _autotest.tag. As far as I can see, the problem is, that only a part of `mH` is filled up with elements in [here](https://github.com/thomas-niehaus/dftbplus/blob/3cdb9035f4114e3e9d526c4845f9a68b09a73678/prog/dftb%2B/lib_timedep/linrespgrad.F90#L1067), while the entire `mH` is used [here](https://github.com/thomas-niehaus/dftbplus/blob/3cdb9035f4114e3e9d526c4845f9a68b09a73678/prog/dftb%2B/lib_timedep/linrespgrad.F90#L1148) Probably it is enough to zero out the superfluous elements, or maybe even better, consistently use only the relevant parts of the matrices.

**Comment by thomas-niehaus** [comment]
This patch solves the problem with NAG on dpdevel. It is true that I should check the dimensions of the various matrices to avoid superfluous memory allocation, but at least all tests pass. I think we are good to merge.

**Comment by aradi** [comment]
OK, perfect, thanks. I'll try to rebase it to the current refactored master.

**Pull request closed without merging**